sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCache.java

```diff
@@ -110,19 +124,14 @@ private Supplier<Mono<? extends AccessToken>> retrieveToken(TokenRequestContext
                         fallback = Mono.just(cache);
                     } else if (cache == null || cache.isExpired()) {
                         // no token to use
-                        if (now.isAfter(nextTokenRefresh)) {
-                            // refresh immediately
-                            tokenRefresh = Mono.defer(tokenSupplier);
-                        } else {
-                            // wait for timeout, then refresh
-                            tokenRefresh = Mono.defer(tokenSupplier)
-                                .delaySubscription(Duration.between(now, nextTokenRefresh));
-                        }
```

Comment:
Are there no cases where a token might be expired but the refresh isn't ready yet? I suspect not but that seems to be the only case that would have made it into this block so I wanted to double check.

Reply:
Yeah, no, the refresh is done via credential, we can invoke it anytime to get a new token.
This seems like a bug, as we currently wait for 30 seconds even if cached token is expired.